### ISSUE TITLE
Desktop - App showed offline status with an active internet connection

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#5337]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
I will add here what @mallenexpensify experienced since I was unable to reproduce on my side: 

> - Was on wifi at hospital, parked on the side of building, reception was poor, slow/unstable, so I gave up on it after 30 mins
> - Started to tether from my phone, which worked a lil better
> - Closed comp
> - Got home, opened comp, wifi connected to home network, I'm able to connect via Slack and visit websites but seeing You appear to be offline. at bottom of both desktop and web, confirmed that test messages I sent did NOT go through.  Gonna refresh then restart to see if that fixes it

## Expected Result:
App should show online if there's an active internet connection. 

## Actual Result:
App showed offline status. 

## Workaround:
Clearing cookies/cache and relogin solved the issue. 

Comment from Matt: 
> Refresh and hard refreshed didn't work in Desktop. clearing local storage and cookies, then signing back in did work. 

## Platform:

Where is this issue occurring?

- Desktop App

**Version Number:** 1.0.98-1

**Reproducible in staging?:** Yes
**Reproducible in production?:** Unknown

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

![image](https://user-images.githubusercontent.com/44479856/133859474-ca29559e-ab3c-485a-94ce-5d2052556b02.png)

**Expensify/Expensify Issue URL:**

**Issue reported by:** @mallenexpensify and also @Santhosh-Sellavel (who reported a similar issue [here](https://github.com/Expensify/App/issues/6059) that we closed, we should compensate @Santhosh-Sellavel if/once we fix this) 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1631820296401500

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @chiragsalian (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by mallenexpensify** [comment]
@chiragsalian can you see if you can magically reproduce? If no, can you add the 'needs reproduction' label and we'll hang tight to see if we can get steps?

**Comment by chiragsalian** [comment]
Yeah I'm not sure how to reproduce just yet. App seems fine to me. Maybe some corrupted localStorage data. Hard to tell atm so i'll apply the label and demote it until we get more reports or reproduction steps.

**Comment by kidroca** [comment]
Did a bit of a research and it seems it's related to this: [Issue 678075: navigator.onLine returns false even though there is internet connectivity](https://bugs.chromium.org/p/chromium/issues/detail?id=678075)
We use this field during connection change events to determine the online/offline state.

There are multiple reports experiencing the issue on a mac - when networks are changed, after sleeping
Users would visit and load different pages successfully, but `navigator.onLine` would be false
The current workaround is to turn wifi off and on again
The odds of experiencing the bug are low, hard to reproduce

Desktop is chromium based so it's plausible it's the same issue
@mallenexpensify what browser did you use when you saw the "offline message" on `web`?

**Comment by mallenexpensify** [comment]
Chrome @kidroca

**Comment by kidroca** [comment]
I think it's something external like the Chrome related issue I've share in my previous comment

I've traced the code that would set and unset our "offline state" and couldn't find anything that might cause this besides the `navigator.onLine` field we rely on. The field is provided by the environment and Expensify.App has no control over it beside being able to read it.

The `net-info` library we use on `web` uses the same field: 
https://github.com/react-native-netinfo/react-native-netinfo/blob/master/src/internal/nativeModule.web.ts#L106
> `const isConnected = navigator.onLine;`

**Comment by kidroca** [comment]
The Chrome issue about `navigator.onLine` being `false` (https://bugs.chromium.org/p/chromium/issues/detail?id=678075) is something that not many people will know they've experienced because it does not do anything on it's own

The browser doesn't do anything special (or anything at all) when `navigator.onLine` is `false`, request are still sent to their destinations and allowed to complete

The issue is only brought up in cases like ours where it's used to do something, most commonly it's used to show a reminder to the user and there's not much of a harm, combined with the fact is hard to reproduce and people are likely to skip reporting it, but in our case we also pause network requests it can results in a bigger problem

---

`navigator.onLine` is not considered a reliable source of connectivity state (though mostly due to false positives). 
Mozilla recommends to use something else in case you need better accuracy

https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine
> Therefore, if you really want to determine the online status of the browser, you should develop additional means for checking. To learn more, see the HTML5 Rocks article, Working Off the Grid.

---

I think we can still use `navigator.onLine` state but combine it with a server check when it changes
E.g. when it changes to `false` make a request against our server, if the server responds switch our app state back to online
When it changes from `false` to `true` make a request to our server to verify we're really connected

**Comment by kidroca** [comment]
I was able to capture this bug and the `navigator.onLine` working incorrectly on my mac 
I've noticed it while I was testing something on the iOS Simulator Safari 

1. Disconnect from a wifi network - nothing happens - `navigator.onLine` remains `true` - you don't see the "You're offline" message 
2. Connect to the same network - `navigator.onLine` is set to `false`, even though you're connected - you see the "You're offline" message

It was pretty consistently recreatable on mobile Safari in the simulator

https://user-images.githubusercontent.com/12156624/140327469-64f925f5-a368-4eaa-96f5-bbf2f73134cc.mov

---

Chrome and Safari are Webkit browsers, and this seems to be a Webkit issue

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days.  eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by mallenexpensify** [comment]
@chiragsalian, @kidroca with the reproducible steps [here](https://github.com/Expensify/App/issues/5337#issuecomment-961007401), do you think this issue can now be worked on by an external contributor?

**Comment by kidroca** [comment]
Sorry @mallenexpensify this showcases it's a bug with the browser an external contributor won't be able to solve it 

To address this ourselves we could move away from using `navigator.onLine` but it would be a very significant change that would surely have to be discussed/planned internally

**Comment by mallenexpensify** [comment]
@kidroca we closed out https://github.com/Expensify/App/issues/6059 is there anything we can/should do with this issue to help it move forward?  I could assign an Expensify engineer to review it to see if should be labeled an internal issue.  

Also, if/when this gets fixed, I added @Santhosh-Sellavel to the 'reported by:' in the OP, let's compensate Santhosh for reporting the issue in #6059 since we closed that.

**Comment by kidroca** [comment]
@mallenexpensify 
I've made a proposal some time ago that can be reviewed: https://github.com/Expensify/App/issues/6059#issuecomment-977296667

Basically whenever we receive an event that our connection state change we double check it by running one request
It can be against a dedicated connection pulse check or against some existing API like get user details, or get tz 
If the server responds to the double check we're obviously still online otherwise we're offline 
When the double check is over we store the result as our current connection state

One more thing that we can add is - when a request fails because it didn't reach the server, we run the double check and if it turns out we're offline we store that as our current connection state

**Comment by mallenexpensify** [comment]
Thanks @kidroca 
Let me get the ball rolling again! @chiragsalian , since you were initially assigned, I'm tossing back to take a look @kidroca 's proposal https://github.com/Expensify/App/issues/6059#issuecomment-977296667  and/or to apply the External label if you think it can be worked on by an external contributor.

**Comment by chiragsalian** [comment]
Hmmm, I do like the ideas but before we go to those ideas could you share the reproduction steps? I tested on chrome and desktop app and `navigator.onLine` gives me false when i disconnect from wifi and gives me true the moment i reconnect to wifi.
I do see in your video you are sort of getting the inverse results which is rather odd. Anyway, I was wondering if there were consistent reproduction steps to confirm the issue? But then again i guess this is pretty rare so perhaps there are no consistent reproduction steps. Either way let me know and yeah i like your idea of making a simple API request to confirm the connection. 

I'm not sure how the proposal will help your scenario though where you are getting an inverted result (from your video) so when you turn off wifi `navigator.onLine` is true and making an API request won't work. Then once you turn wifi if `navigator.onLine` is false are we going to still make an API request to change the value of isConnected. It'll be odd making requests when `navigator.onLine` is false. Let me know your thoughts.

**Comment by chiragsalian** [comment]
@kidroca, friendly bump on my previous comment when you get the chance. (mostly posting so that melvin doesn't consider this overdue)

**Comment by kidroca** [comment]
Thanks for the bump, I've seen and forgotten about this one

I don't think there are solid reproduction steps for "offline status with an active internet connection" 
It seems to be this Chrome/Webkit issue where `navigator.onLine` would tell `false` even though you're online: https://github.com/Expensify/App/issues/5337#issuecomment-943543012
It seems it's affecting only macs and isn't happening regularly - So far I can consistently recreate it in simulators, but it happened only one time in actual Safari outside the simulator

> I'm not sure how the proposal will help your scenario though where you are getting an inverted result

Yep, it'll probably not work because:
1. `navigator.onLine` goes to `true` 
2. We make a double check and we see it fail - we set app state as offline
3. Later `navigator.onLine` goes to `false` 
4. We see we've already set the app state as offline and we do nothing

And there are other problems as well like:
1. There's a network connection, but there's no internet (like when you kill mobile data on your tethered connection)
2. `navigator.onLine` is `true`
3. Our request are failing, not reaching the backend - we set app state as offline
3. Later the app really becomes online but since `onLine` is already `true` no event is raised
4. Nothing ever causes us to re-check network state and we keep app state to offline
   - in the end we're left in a incorrect state

I think these examples demonstrate we don't necessarily have to rely on `navigation.onLine` and while the Webkit bug is rare, the other one where `navigator.onLine` will say `true` simply because you're connected to a network, but not necessarily that there's an internet connection, warrant taking additional measures 

---

My original idea was to ditch `navigator.onLine`
This would still allows us to track online/offline state:

1. The app is always considered online at launch
2. If a request fails due to timeout or no response we set offline mode
3. We continue to allow new requests, but persist every failed request to be retired later
4. If a request succeeds we set online mode and we execute the list of persisted requests
5. Optional: While we're offline, a `pulse` request is made every minute or two - to trigger 3) and 4). Pules should not be persisted

(3) Is similar to how App currently works - it pushes requests to a list that will be retried once we're back online
The main difference is we attempt the request regardless of connection state, and only set it to be retired if it fails
And  4) is the same thing that happens when we receive an event that we're online 
5) might be unnecessary - almost all interactions with the app would trigger some kind of request so it would not be necessary to poll e.g. opening a chat would trigger an `UpdateLastRead` request, focusing the app would refetch LHN reports

Another benefit of point 3) is that we start with the latest request
If you were offline, on a spotty connection, and have accumulated over 20+ requests
When we're only making your latest request and it succeeds you'll see fresh information on the report you're currently observing, and this will still trigger a chain reaction to fetch the rest of the requests

In comparison what happens now is as soon as (something tells us) we're back online we call all the requests
and they all start in parallel, but only about 5 can run at a time - the rest would get stalled by browser/env
Usually they are handled in FIFO, so if you're back online with 20+ request pending, on a dodgy connection, starting with the oldest requests would be the slowest way to bring relevant information to the user

**Comment by chiragsalian** [comment]
Your solutions are interesting. I think point 5 of your solution is needed because for a user there should be some visual indicator to know if their app is online or not otherwise they would be staring at the screen unsure if things are fine or not.

Since I haven't come across this problem I'm unsure how widespread and if this is a good time to make the suggested changes. But I'm not against it if it can help make our product better. I'll loop in @marcaaron and @roryabraham for thoughts since they might be more familiar with `navigator.onLine` and if they feel like it's a good time to ditch `navigator.onLine` to track online/offline.

**Comment by marcaaron** [comment]
I haven't read through this entire thread, but might be better to have with the `net-info` maintainers to see if they have ideas on how to solve it. My preference would be to not ditch this library since it works quite well for most of our cases.

Keep in mind there are known issues with it not working well with Simulator see [here](https://github.com/react-native-netinfo/react-native-netinfo#issues-with-the-ios-simulator).

**Comment by kidroca** [comment]
As per `net-info` documentation https://github.com/react-native-netinfo/react-native-netinfo#browser-compatilibity
> The web implementation heavily depends on the [Network Information API](https://developer.mozilla.org/en-US/docs/Web/API/Network_Information_API) which is still an is an experimental technology and thus it's not supported in every browser. If this API is not available the library will safely fallback to the old [onLine](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/onLine) property and return basic connection information.

So it's unlikely they'll do anything more than that 

What I see in out code is that we treat `isConnected` as an indication of an internet access while it's only an indication of having a network connection: https://github.com/react-native-netinfo/react-native-netinfo#netinfostate and there's actually an `isInternetReachable` key that maybe we should use

https://github.com/Expensify/App/blob/838704a7a520065407bd8a731925e155648bb99d/src/libs/NetworkConnection.js#L57-L63

Also there's [configuration](https://github.com/react-native-netinfo/react-native-netinfo#netinfoconfiguration) where we can provide a `reachabilityUrl` and `reachabilityTest`. This is similar to what was proposed earlier

So maybe we should first try to
- use `isInternetReachable` and see if it works
- configure our own `reachabilityTest` (if that's going to make it work better, or on all platforms)

**Comment by kidroca** [comment]
I've inspected how `reachabilityTest` test works.
It will be performed when a network change occurs, then if it fails it will be polled until it succeeds (the `reachabilityLongTimeout` configures this)

As far as I see nothing would trigger `reachabilityTest` if our connection simply loses internet access
But we can force a recheck by calling `NetInfo.fetch()`
So if we stick with `NetInfo` we should do that when a request of our own fails without a response 

I see a number of `isInternetReachable` issues open dating back to more than a year ago:  https://github.com/react-native-netinfo/react-native-netinfo/issues?q=is%3Aissue+isInternetReachable+is%3Aopen

`reachabilityUrl` it's not used on all the platforms 
> Only used on platforms which do not supply internet reachability natively

So far this seems to be iOS only, and that's where some issue are

---

If we're only interested in internet reachability the latest proposal [here](https://github.com/Expensify/App/issues/5337#issuecomment-987268111) should be a more reliable way to achieve that (on all platforms)

As far as I can see we're not interested in anything more ATM

**Comment by MelvinBot** [comment]
@chiragsalian Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by kidroca** [comment]
We decided to use this ticket to address online/offline state not being correctly depicted by App: https://github.com/Expensify/App/issues/5337#issuecomment-985127489

That's both:
- when the app is falsely assuming we're **offline** (the rare case)
- when the app is falsely assuming we're **online** (#6059)
    - this bug can be recreated consistently

The comment about our code not working correctly is regarding the more common case of #6059:
> What I see in out code is that we treat `isConnected` as an indication of an internet access while it's only an indication of having a network connection: https://github.com/react-native-netinfo/react-native-netinfo#netinfostate and there's actually an `isInternetReachable` key that maybe we should use
> 
> https://github.com/Expensify/App/blob/838704a7a520065407bd8a731925e155648bb99d/src/libs/NetworkConnection.js#L57-L63

We can at least address that by switching to `isInternetReachable` and optionally adding our own `reachabilityTest`

**Comment by chiragsalian** [comment]
Cool, I didn't see https://github.com/Expensify/App/issues/6059 before. I have an iPhone so I cannot reproduce that issue since once I turn off cellular it also turns off personal hotspot. Seems like quite the edge case even on android devices. 

I'm not opposed to adding `isInternetReachable` and our own `reachabilityTest` since it does sound like its a better indication of internet access than network connectivity that you mentioned. I'm just wondering if its worth the effort since it sounds like quite the edge case on android devices. Any second opinion on this @marcaaron or @roryabraham?

**Comment by kidroca** [comment]
That's really a rare case (tethering), but it was only used to showcase the real issue - being connected to a network with no internet 
It has happened to me on public WIFIs (on airports) where you don't have internet access until you subscribe or use some code
I imagine moving through a weak reception can lead to the same issue, fooling the system by being barely connected

Adding `isInternetReachable` would be no effort at all, unless it turns out it doesn't work

**Comment by roryabraham** [comment]
It seems to me like using the `isInternetReachable` state would be an easy potential solution to https://github.com/Expensify/App/issues/6059, but I don't see how it would address this issue. Adding our own `reachabilityTest` also seems like it might be more complexity than it's worth for these edge cases. Per their documentation,  it should really only be necessary on platforms that "don't supply internet reachability natively". Does that apply to any of our platforms, or is this just a workaround for the [rare chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=678075)?

**Comment by chiragsalian** [comment]
Thanks for those thoughts Rory. I feel like the issue doesn't feel like its worth the ROI at the moment and so i propose that we close this issue for now to work on tasks of higher importance. If we ever feel like this task is higher up on our ROI then we come back to it. Thoughts?

**Comment by mallenexpensify** [comment]
My concern with closing this is we have two issues here (as mentioned by @kidroca above)

- when the app is falsely assuming we're **offline** (the rare case)
- when the app is falsely assuming we're **online** (#6059)
    - this bug can be recreated consistently
    
#6059 has already been closed cuz it was considered a dupe of this one.  If we now close this, we will not have _solved_ either.   If tethering is considered rare and #6059 can be reproduced consistently, should we close this and reopen #6059 so we can try to fix the more common bug?   That sounds like the best path to me

**Comment by roryabraham** [comment]
> If tethering is considered rare and #6059 can be reproduced consistently, should we close this and reopen #6059 so we can try to fix the more common bug? That sounds like the best path to me

That makes sense to me.

**Comment by chiragsalian** [comment]
> If tethering is considered rare and #6059 can be reproduced consistently, should we close this and reopen #6059 so we can try to fix the more common bug? That sounds like the best path to me

Sure I don't mind tackling this issue. The issue is only reproducible on android phones and you cannot reproduce it on iPhones so i can't really test on my end. So yeah @kidroca if you would like feel free to go ahead and implement `isInternetReachable` and let us know if it works any better while testing with an android phone.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by mallenexpensify** [comment]
I reopened the other issue, created a new Upwork job for it and doubled the price to $500.  

I proposed we close this issue, @chiragsalian do you agree?

**Comment by chiragsalian** [comment]
Sure, closing this one out.
